dlt/destinations/impl/filesystem/sql_client.py

```diff
@@ -0,0 +1,250 @@
+from typing import Any, Iterator, AnyStr, List, cast, TYPE_CHECKING, Dict
+
+import os
+import re
+
+import dlt
+
+import duckdb
+
+import sqlglot
+import sqlglot.expressions as exp
+from dlt.common import logger
+
+from contextlib import contextmanager
+
+from dlt.common.destination.reference import DBApiCursor
+from dlt.common.destination.typing import PreparedTableSchema
+
+from dlt.destinations.sql_client import raise_database_error
+
+from dlt.destinations.impl.duckdb.sql_client import DuckDbSqlClient
+from dlt.destinations.impl.duckdb.factory import duckdb as duckdb_factory, DuckDbCredentials
+from dlt.common.configuration.specs import (
+    AwsCredentials,
+    AzureServicePrincipalCredentialsWithoutDefaults,
+    AzureCredentialsWithoutDefaults,
+)
+
+SUPPORTED_PROTOCOLS = ["gs", "gcs", "s3", "file", "memory", "az", "abfss"]
+
+if TYPE_CHECKING:
+    from dlt.destinations.impl.filesystem.filesystem import FilesystemClient
+else:
+    FilesystemClient = Any
+
+
+class FilesystemSqlClient(DuckDbSqlClient):
+    def __init__(
+        self,
+        fs_client: FilesystemClient,
+        dataset_name: str = None,
+        duckdb_connection: duckdb.DuckDBPyConnection = None,
```

Comment:
I changed it so you can now hand over a credentials object to the filesystem duckdb client. I find it less convenient because you need to wrap the connection in a credentials object, but I suppose it is better like that.